src/Microsoft.Data/ArrowStringColumn.cs

```diff
@@ -0,0 +1,348 @@
+            }
+            else
+            {
+                byte[] bytes = _encoding.GetBytes(value);
+                MutableDataFrameBuffer<byte> mutableDataBuffer = _dataBuffers[_dataBuffers.Count - 1] as MutableDataFrameBuffer<byte>;
+                if (mutableDataBuffer.Length == DataFrameBuffer<byte>.MaxCapacity)
+                {
+                    mutableDataBuffer = new MutableDataFrameBuffer<byte>();
+                    _dataBuffers.Add(mutableDataBuffer);
+                    _nullBitMapBuffers.Add(new MutableDataFrameBuffer<byte>());
+                    var offsetBuffer = new MutableDataFrameBuffer<int>();
+                    _offsetsBuffers.Add(offsetBuffer);
+                    offsetBuffer.Append(0);
+                }
+                mutableDataBuffer.EnsureCapacity(bytes.Length);
+                bytes.AsMemory().CopyTo(mutableDataBuffer.Memory.Slice(mutableDataBuffer.Length));
+                mutableDataBuffer.Length += bytes.Length;
+                mutableOffsetsBuffer.Append(mutableOffsetsBuffer[mutableOffsetsBuffer.Length - 1] + bytes.Length);
+            }
+            SetValidityBit(Length - 1, value == null ? true : false);
+        }
+
+        private int GetBufferIndexContainingRowIndex(ref long rowIndex)
+        {
+            if (rowIndex >= Length)
+            {
+                throw new ArgumentOutOfRangeException(Strings.ColumnIndexOutOfRange, nameof(rowIndex));
+            }
+
+            // Since the strings here could be of variable length, scan linearly
+            int curArrayIndex = 0;
+            int numBuffers = _offsetsBuffers.Count;
+            while (curArrayIndex < numBuffers && rowIndex > _offsetsBuffers[curArrayIndex].Length - 1)
+            {
+                rowIndex -= _offsetsBuffers[curArrayIndex].Length - 1;
+                curArrayIndex++;
+            }
+            return curArrayIndex;
+        }
+
+        private ReadOnlySpan<byte> GetBytes(long index)
+        {
+            int offsetsBufferIndex = GetBufferIndexContainingRowIndex(ref index);
+            int currentOffset = _offsetsBuffers[offsetsBufferIndex].ReadOnlySpan[(int)index];
+            int nextOffset = _offsetsBuffers[offsetsBufferIndex].ReadOnlySpan[(int)index + 1];
+            int numberOfBytes = nextOffset - currentOffset;
+            return _dataBuffers[offsetsBufferIndex].ReadOnlySpan.Slice(currentOffset, numberOfBytes);
+        }
+
+        protected override object GetValue(long rowIndex)
+        {
+            var bytes = GetBytes(rowIndex);
+            unsafe
+            {
+                fixed (byte* data = &MemoryMarshal.GetReference(bytes))
+                    return _encoding.GetString(data, bytes.Length);
+            }
+        }
+
+        protected override object GetValue(long startIndex, int length)
+        {
+            var ret = new List<string>();
+            while (ret.Count < length)
+            {
+                ret.Add((string)GetValue((int)startIndex));
```

Comment:
Don't you need to increment startIndex here?

BTW - do we have tests for this method?

Reply:
This method could just `return this[startIndex, length];` right?

---
In reply to: [307378768](https://github.com/dotnet/corefxlab/pull/2701#discussion_r307378768) [](ancestors = 307378768)

Reply:
Good catch on the startIndex. I added a test for this.

No, this is the implementation of `this[startIndex, length]`.